File: 238-product-of-array-except-self/238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] result = new int[nums.length];
        
        Arrays.fill(result, 1);
        
        int curr = 0;
        
        while (curr < result.length) {
            for (int i = 0; i < nums.length; i++) {
                if (curr == i) {
                    continue;
                } else {
                    result[curr] *= nums[i];
                }
            }
            curr++;
        }
        return result;
    }
}

// this solution works, it's just less efficient -> doesn't use the division operator but it does run in O(n^2) time)

File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        
        length = len(nums)
        product = 1
        zeroesCount = 0
        
        for i in nums:
            if i == 0:
                zeroesCount += 1
            else:
                product *= i
                
        if zeroesCount > 1:
            return [0] * length
        elif zeroesCount == 1:
            for i in range(length):
                if nums[i] != 0:
                    nums[i] = 0;
                else:
                    nums[i] = product
        else: 
            for i in range(length):
                nums[i] = product // nums[i]
        
        return nums